File: project/drivers/driver/r/R/sql.R
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

# Return the new terminator string if `chunk` is a `SET TERM <terminator>`
# client directive, else NULL. Leading full-line `--` comments and blank lines
# are ignored when matching, so a directive may be preceded by comment lines in
# the same chunk.
chunk_set_term <- function(chunk) {
  lines <- strsplit(chunk, "\n", fixed = TRUE)[[1]]
  meaningful <- character(0)
  for (line in lines) {
    stripped <- trimws(line)
    if (!nzchar(stripped) || startsWith(stripped, "--")) next
    meaningful <- c(meaningful, stripped)
  }
  if (length(meaningful) == 0) return(NULL)
  joined <- paste(meaningful, collapse = " ")
  m <- regmatches(joined, regexec("^[Ss][Ee][Tt]\\s+[Tt][Ee][Rr][Mm]\\s+(\\S.*?)\\s*$", joined))[[1]]
  if (length(m) < 2) return(NULL)
  rest <- trimws(m[2])
  if (!nzchar(rest)) return(NULL)
  rest
}

# Split SQL into top-level statements on the active terminator.
#
# Quote-aware (single/double quotes) and `--` line-comment aware. Honors the
# `SET TERM <terminator>` client directive: the
# directive changes the active terminator and is consumed -- it is not emitted as
# a statement and is not counted. This lets procedural bodies contain inner `;`
# between `SET TERM ^` and the restoring `SET TERM ;^`. With no `SET TERM`
# directive present, this reduces to a plain quote-aware top-level `;` split.
split_top_level_statements <- function(sql) {
  statements <- list()
  term <- ";"
  lines <- strsplit(sql, "\n", fixed = TRUE)[[1]]
  if (length(lines) == 0) return(character())
  in_single <- FALSE
  in_double <- FALSE
  buf <- character(0)

  append_piece <- function(piece) {
    if (nzchar(piece)) {
      buf[[length(buf) + 1]] <<- piece
    }
  }

  flush <- function() {
    chunk <- trimws(paste(buf, collapse = ""))
    buf <<- character(0)
    if (!nzchar(chunk)) return(invisible(NULL))
    new_term <- chunk_set_term(chunk)
    if (!is.null(new_term)) {
      term <<- new_term
      return(invisible(NULL))
    }
    statements[[length(statements) + 1]] <<- chunk
    invisible(NULL)
  }

  for (line_index in seq_along(lines)) {
    line <- lines[[line_index]]
    n <- nchar(line, type = "chars", allowNA = FALSE, keepNA = FALSE)
    i <- 1
    segment_start <- 1
    while (i <= n) {
      ch <- substr(line, i, i)
      if (!in_single && !in_double && ch == "-" && i + 1 <= n && substr(line, i + 1, i + 1) == "-") {
        append_piece(substr(line, segment_start, n))
        segment_start <- n + 1
        i <- n + 1
        break
      }
      if (ch == "'" && !in_double) {
        in_single <- !in_single
        i <- i + 1
        next
      }
      if (ch == '"' && !in_single) {
        in_double <- !in_double
        i <- i + 1
        next
      }
      term_len <- nchar(term, type = "chars", allowNA = FALSE, keepNA = FALSE)
      if (!in_single && !in_double && term_len > 0 &&
          i + term_len - 1 <= n &&
          identical(substr(line, i, i + term_len - 1), term)) {
        if (i > segment_start) {
          append_piece(substr(line, segment_start, i - 1))
        }
        flush()
        i <- i + term_len
        segment_start <- i
        next
      }
      i <- i + 1
    }
    if (segment_start <= n) {
      append_piece(substr(line, segment_start, n))
    }
    if (line_index < length(lines) && length(buf) > 0) {
      append_piece("\n")
    }
  }
  flush()
  vapply(statements, identity, character(1))
}

sb_normalize <- function(sql, params = NULL) {
  if (is.null(params) || length(params) == 0) {
    return(list(sql = sql, params = list()))
  }
  if (!is.null(names(params)) && any(names(params) != "")) {
    return(rewrite_named(sql, params))
  }
  if (grepl("\\?", sql, fixed = FALSE)) {
    return(rewrite_positional(sql, params))
  }
  list(sql = sql, params = params)
}

has_named_params <- function(sql) {
  in_string <- FALSE
  chars <- strsplit(sql, "", fixed = TRUE)[[1]]
  for (i in seq_len(length(chars) - 1)) {
    ch <- chars[i]
    if (ch == "'") {
      in_string <- !in_string
      next
    }
    if (!in_string && (ch == ":" || ch == "@")) {
      next_ch <- chars[i + 1]
      if (grepl("[A-Za-z_]", next_ch)) return(TRUE)
    }
  }
  FALSE
}

rewrite_named <- function(sql, params) {
  if (!has_named_params(sql)) stop("named parameters provided but query has no named placeholders")
  lookup <- list()
  for (name in names(params)) {
    key <- sub("^[@:]", "", name)
    lookup[[key]] <- params[[name]]
  }
  out <- ""
  ordered <- list()
  in_string <- FALSE
  chars <- strsplit(sql, "", fixed = TRUE)[[1]]
  i <- 1
  while (i <= length(chars)) {
    ch <- chars[i]
    if (ch == "'") {
      in_string <- !in_string
      out <- paste0(out, ch)
      i <- i + 1
      next
    }
    if (!in_string && (ch == ":" || ch == "@") && i + 1 <= length(chars) && grepl("[A-Za-z_]", chars[i + 1])) {
      j <- i + 1
      while (j <= length(chars) && grepl("[A-Za-z0-9_]", chars[j])) {
        j <- j + 1
      }
      name <- paste(chars[(i + 1):(j - 1)], collapse = "")
      if (is.null(lookup[[name]])) stop(paste0("missing named parameter: ", name))
      ordered[[length(ordered) + 1]] <- lookup[[name]]
      out <- paste0(out, "$", length(ordered))
      i <- j
      next
    }
    out <- paste0(out, ch)
    i <- i + 1
  }
  list(sql = out, params = ordered)
}

rewrite_positional <- function(sql, params) {
  out <- ""
  ordered <- list()
  in_string <- FALSE
  chars <- strsplit(sql, "", fixed = TRUE)[[1]]
  i <- 1
  idx <- 1
  while (i <= length(chars)) {
    ch <- chars[i]
    if (ch == "'") {
      in_string <- !in_string
      out <- paste0(out, ch)
      i <- i + 1
      next
    }
    if (!in_string && ch == "?") {
      if (idx > length(params)) stop("not enough parameters")
      ordered[[length(ordered) + 1]] <- params[[idx]]
      idx <- idx + 1
      out <- paste0(out, "$", length(ordered))
      i <- i + 1
      next
    }
    out <- paste0(out, ch)
    i <- i + 1
  }
  if (idx <= length(params)) stop("too many parameters")
  list(sql = out, params = ordered)
}

File: project/drivers/driver/php/src/Sql.php
<?php
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

namespace ScratchBird\PDO;

final class Sql
{
    public static function normalize(string $sql, array $params): array
    {
        if (empty($params)) {
            return ['sql' => $sql, 'params' => []];
        }

        if (self::hasNamedParameters($sql)) {
            return self::rewriteNamed($sql, $params);
        }

        if (str_contains($sql, '?')) {
            return self::rewritePositional($sql, $params);
        }

        return ['sql' => $sql, 'params' => array_values($params)];
    }

    public static function normalizeCallable(string $sql, array $params): array
    {
        return self::normalize(self::normalizeCallableSql($sql), $params);
    }

    /**
     * @param array<int, mixed> $params
     * @return array<int, array{sql: string, params: array<int, mixed>}>|null
     */
    public static function splitExecutableStatements(string $sql, array $params): ?array
    {
        if (self::isSingleProceduralDefinition($sql)) {
            return null;
        }
        $statements = self::splitTopLevelStatements($sql);
        if (count($statements) <= 1) {
            return null;
        }
        $out = [];
        foreach ($statements as $statement) {
            $out[] = self::remapStatementParams($statement, $params);
        }
        return $out;
    }

    public static function normalizeCallableSql(string $sql): string
    {
        $trimmed = trim($sql);
        if (!str_starts_with($trimmed, '{') || !str_ends_with($trimmed, '}')) {
            return $sql;
        }
        $inner = trim(substr($trimmed, 1, -1));
        if ($inner === '') {
            return $sql;
        }

        if (preg_match('/^\?\s*=\s*call\s+([\s\S]+)$/i', $inner, $matches) === 1) {
            $parsed = self::parseCallableInvocation(trim($matches[1]));
            if ($parsed === null) {
                throw new \InvalidArgumentException('invalid JDBC escape call syntax');
            }
            [$routine, $args, $hasParens] = $parsed;
            $callArgs = $hasParens ? $args : '';
            return sprintf('select %s(%s) as return_value', $routine, $callArgs);
        }

        if (preg_match('/^call\s+([\s\S]+)$/i', $inner, $matches) === 1) {
            $parsed = self::parseCallableInvocation(trim($matches[1]));
            if ($parsed === null) {
                throw new \InvalidArgumentException('invalid JDBC escape call syntax');
            }
            [$routine, $args, $hasParens] = $parsed;
            if ($hasParens) {
                return sprintf('call %s(%s)', $routine, $args);
            }
            return sprintf('call %s', $routine);
        }

        return $sql;
    }

    private static function hasNamedParameters(string $sql): bool
    {
        $len = strlen($sql);
        $inString = false;
        for ($i = 0; $i + 1 < $len; $i++) {
            $ch = $sql[$i];
            if ($ch === "'") {
                $inString = !$inString;
                continue;
            }
            if ($inString) {
                continue;
            }
            if (self::isNamedParameterStart($sql, $i)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Split SQL into top-level statements on the active terminator.
     *
     * Quote-aware (single/double quotes) and `--` line-comment aware. Honors the
     * `SET TERM <terminator>` client directive:
     * the directive changes the active terminator and is consumed -- it is not
     * emitted as a statement and is not counted in statement indexing. This lets
     * procedural bodies contain inner `;` between `SET TERM ^` and the restoring
     * `SET TERM ;^`.
     *
     * With no `SET TERM` directive present, the behavior reduces to a plain
     * quote-aware top-level `;` split, so existing scripts and statement indices
     * are unchanged.
     *
     * @return array<int, string>
     */
    private static function splitTopLevelStatements(string $sql): array
    {
        $statements = [];
        $term = ';';
        $current = '';
        $len = strlen($sql);
        $inString = false;
        $inDouble = false;

        $flush = static function () use (&$current, &$term, &$statements): void {
            $chunk = trim($current);
            $current = '';
            if ($chunk === '') {
                return;
            }
            $newTerm = self::chunkSetTerm($chunk);
            if ($newTerm !== null) {
                $term = $newTerm;
                return;
            }
            $statements[] = $chunk;
        };

        for ($i = 0; $i < $len;) {
            $ch = $sql[$i];
            if (!$inString && !$inDouble && $ch === '-' && $i + 1 < $len && $sql[$i + 1] === '-') {
                // `--` line comment: copy verbatim to end of line without scanning
                // for the terminator or quotes inside it.
                $eol = strpos($sql, "\n", $i);
                if ($eol === false) {
                    $eol = $len;
                }
                $current .= substr($sql, $i, $eol - $i);
                $i = $eol;
                continue;
            }
            if ($ch === "'" && !$inDouble) {
                $inString = !$inString;
                $current .= $ch;
                $i++;
                continue;
            }
            if ($ch === '"' && !$inString) {
                $inDouble = !$inDouble;
                $current .= $ch;
                $i++;
                continue;
            }
            if (!$inString && !$inDouble && $term !== '' && substr_compare($sql, $term, $i, strlen($term)) === 0) {
                $matchedLen = strlen($term); // capture before flush(), which may change $term
                $flush();
                $i += $matchedLen;
                continue;
            }
            $current .= $ch;
            $i++;
        }
        $flush();
        return $statements;
    }

    /**
     * Return the new terminator if `$chunk` is a `SET TERM <terminator>` client
     * directive, else `null`. Leading full-line `--` comments and blank lines are
     * ignored when matching, so a directive may be preceded by comment lines.
     */
    private static function chunkSetTerm(string $chunk): ?string
    {
        $meaningful = [];
        foreach (preg_split('/\r\n|\r|\n/', $chunk) as $line) {
            $stripped = trim($line);
            if ($stripped === '' || str_starts_with($stripped, '--')) {
                continue;
            }
            $meaningful[] = $stripped;
        }
        if ($meaningful === []) {
            return null;
        }
        if (preg_match('/^set\s+term\s+(\S.*?)\s*$/i', implode(' ', $meaningful), $matches) === 1) {
            return trim($matches[1]);
        }
        return null;
    }

    private static function isSingleProceduralDefinition(string $sql): bool
    {
        $trimmed = self::trimLeadingLineComments($sql);
        return preg_match(
            '/^create\s+(?:or\s+replace\s+)?(?:trigger|function|procedure)\b/i',
            $trimmed
        ) === 1;
    }

    private static function trimLeadingLineComments(string $sql): string
    {
        $lines = preg_split('/\r\n|\r|\n/', $sql);
        if ($lines === false) {
            return ltrim($sql);
        }
        $offset = 0;
        $count = count($lines);
        while ($offset < $count) {
            $line = ltrim($lines[$offset]);
            if ($line === '' || str_starts_with($line, '--')) {
                $offset++;
                continue;
            }
            break;
        }
        return ltrim(implode("\n", array_slice($lines, $offset)));
    }

    private static function rewriteNamed(string $sql, array $params): array
    {
        $lookup = [];
        foreach ($params as $key => $value) {
            if (is_string($key)) {
                $lookup[ltrim($key, '@:')] = $value;
            }
        }
        $ordered = [];
        $out = '';
        $len = strlen($sql);
        $inString = false;
        for ($i = 0; $i < $len;) {
            $ch = $sql[$i];
            if ($ch === "'") {
                $inString = !$inString;
                $out .= $ch;
                $i++;
                continue;
            }
            if (!$inString && self::isNamedParameterStart($sql, $i)) {
                $j = $i + 1;
                while ($j < $len && (ctype_alnum($sql[$j]) || $sql[$j] === '_')) {
                    $j++;
                }
                $name = substr($sql, $i + 1, $j - $i - 1);
                if (!array_key_exists($name, $lookup)) {
                    throw new \InvalidArgumentException("missing named parameter: {$name}");
                }
                $ordered[] = $lookup[$name];
                $out .= '$' . count($ordered);
                $i = $j;
                continue;
            }
            $out .= $ch;
            $i++;
        }
        return ['sql' => $out, 'params' => $ordered];
    }

    private static function rewritePositional(string $sql, array $params): array
    {
        $ordered = [];
        $out = '';
        $len = strlen($sql);
        $inString = false;
        $index = 0;
        for ($i = 0; $i < $len;) {
            $ch = $sql[$i];
            if ($ch === "'") {
                $inString = !$inString;
                $out .= $ch;
                $i++;
                continue;
            }
            if (!$inString && $ch === '?') {
                if (!array_key_exists($index, $params)) {
                    throw new \InvalidArgumentException('not enough parameters');
                }
                $ordered[] = $params[$index];
                $index++;
                $out .= '$' . count($ordered);
                $i++;
                continue;
            }
            $out .= $ch;
            $i++;
        }
        if ($index < count($params)) {
            throw new \InvalidArgumentException('too many parameters');
        }
        return ['sql' => $out, 'params' => $ordered];
    }

    /**
     * @param array<int, mixed> $params
     * @return array{sql: string, params: array<int, mixed>}
     */
    private static function remapStatementParams(string $sql, array $params): array
    {
        if ($params === []) {
            return ['sql' => $sql, 'params' => []];
        }

        $len = strlen($sql);
        $inString = false;
        $inDouble = false;
        $out = '';
        $remap = [];
        $orderedIndexes = [];
        for ($i = 0; $i < $len;) {
            $ch = $sql[$i];
            if ($ch === "'" && !$inDouble) {
                $inString = !$inString;
                $out .= $ch;
                $i++;
                continue;
            }
            if ($ch === '"' && !$inString) {
                $inDouble = !$inDouble;
                $out .= $ch;
                $i++;
                continue;
            }
            if (!$inString && !$inDouble && $ch === '$' && $i + 1 < $len && ctype_digit($sql[$i + 1])) {
                $j = $i + 1;
                while ($j < $len && ctype_digit($sql[$j])) {
                    $j++;
                }
                $originalIndex = (int) substr($sql, $i + 1, $j - $i - 1);
                if (!isset($remap[$originalIndex])) {
                    $remap[$originalIndex] = count($orderedIndexes) + 1;
                    $orderedIndexes[] = $originalIndex;
                }
                $out .= '$' . $remap[$originalIndex];
                $i = $j;
                continue;
            }
            $out .= $ch;
            $i++;
        }

        $ordered = [];
        foreach ($orderedIndexes as $originalIndex) {
            if ($originalIndex < 1 || !array_key_exists($originalIndex - 1, $params)) {
                throw new \InvalidArgumentException('parameter count mismatch');
            }
            $ordered[] = $params[$originalIndex - 1];
        }
        return ['sql' => $out, 'params' => $ordered];
    }

    private static function isNamedParameterStart(string $sql, int $index): bool
    {
        $len = strlen($sql);
        if ($index < 0 || $index + 1 >= $len) {
            return false;
        }

        $marker = $sql[$index];
        if ($marker !== ':' && $marker !== '@') {
            return false;
        }
        if (!ctype_alpha($sql[$index + 1])) {
            return false;
        }
        if ($marker === ':' && $index > 0 && $sql[$index - 1] === ':') {
            return false;
        }
        return true;
    }

    /**
     * @return array{0: string, 1: string, 2: bool}|null
     */
    private static function parseCallableInvocation(string $text): ?array
    {
        $openParen = strpos($text, '(');
        if ($openParen === false) {
            $routine = trim($text);
            if ($routine === '') {
                return null;
            }
            return [$routine, '', false];
        }

        $inSingle = false;
        $inDouble = false;
        $depth = 0;
        $closeParen = -1;
        $len = strlen($text);
        for ($i = $openParen; $i < $len; $i++) {
            $ch = $text[$i];
            if ($ch === "'" && !$inDouble) {
                $inSingle = !$inSingle;
                continue;
            }
            if ($ch === '"' && !$inSingle) {
                $inDouble = !$inDouble;
                continue;
            }
            if ($inSingle || $inDouble) {
                continue;
            }
            if ($ch === '(') {
                $depth++;
                continue;
            }
            if ($ch === ')') {
                $depth--;
                if ($depth === 0) {
                    $closeParen = $i;
                    break;
                }
            }
        }

        if ($closeParen < 0) {
            return null;
        }
        $routine = trim(substr($text, 0, $openParen));
        if ($routine === '') {
            return null;
        }
        $trailing = trim(substr($text, $closeParen + 1));
        if ($trailing !== '') {
            return null;
        }
        $args = trim(substr($text, $openParen + 1, $closeParen - $openParen - 1));
        return [$routine, $args, true];
    }
}
